### ISSUE TITLE
--warn-unreachable and isinstance leads to false positive about method resolution order

### ISSUE THREAD

**Issue by kstauffer** [python/mypy#15038]
**Bug Report**

A small V shaped class hierarchy, a variable annotated as an instance of the supertype, and a few `assert isinstance` to progressively narrow the type gets a false positive error message from mypy that no such subclass could exist. The code works fine at runtime, though.

**To Reproduce**

https://gist.github.com/mypy-play/a653520575a1e3279bfcec2f64c4a128

Requires --warn-unreachable:
https://mypy-play.net/?mypy=latest&python=3.11&flags=warn-unreachable&gist=a653520575a1e3279bfcec2f64c4a128

```python
class A: pass
class B(A): pass
class C:pass
class D(B, C):pass

m: A = D()
assert isinstance(m, C)  # remove this line, or move it to after the following one, and the error goes away.
assert isinstance(m, D)  # error: Subclass of "A", "C", and "D" cannot exist: would have inconsistent method resolution order  [unreachable]
```

**Actual Behavior**

```
$ mypy --warn-unreachable b.py
b.py:8: error: Subclass of "A", "C", and "D" cannot exist: would have inconsistent method resolution order  [unreachable]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.2.0 (compiled: yes)
- Mypy command-line flags: --warn-unreachable
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.11.0rc1

**Comment by erictraut** [comment]
Why do you think this is a false positive? It doesn't appear to be possible to define a class that is a subclass of `A`, `C` and `D` without getting a runtime error due to inconsistent method resolution order.

```python
class Foo1(A, C, D): ...
class Foo2(D, A, C): ...
```

So I think mypy is correct here.

**Comment by kstauffer** [comment]
Run the reproducer. The asserts pass at runtime. `m` is a D. So it is also a C, a B, and an A. mypy’s logic surrounding isinstance is wrong in this case.

Your counter example explicitly constructs a class with an inconsistent MRO. That isn’t the case in the example shown here in this issue.
